lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,873 @@
+        CurrentCacheBlockSize %= CacheBlockSize;
+        CurrentCacheBlock.clear();
+        if (CurrentCacheBlockSize > 0)
+          CurrentCacheBlock.push_back(FieldNum);
+      }
+    }
+  }
+  WCP += calculateWCPWithinCacheBlock(&CurrentCacheBlock);
+  return WCP;
+}
+
+// Wrapper function to calculate potential WCP if put field FieldNum at
+// the back of NewOrder list
+double FieldReorderAnalyzer::estimateWCPAtBack(FieldNumType FieldNum) {
+  NewOrder.push_back(FieldNum);
+  auto WCP = getWCP();
+  NewOrder.pop_back();
+  return WCP;
+}
+
+// Wrapper function to calculate potential WCP if put field FieldNum at
+// the front of NewOrder list
+double FieldReorderAnalyzer::estimateWCPAtFront(FieldNumType FieldNum) {
+  NewOrder.push_front(FieldNum);
+  auto WCP = getWCP();
+  NewOrder.pop_front();
+  return WCP;
+}
+
+// Main function to make suggestions.
+// The algorithm is to find a pair of fields that has the largest CP value of
+// all fields and put them into NewOrder as seeds (already done in constructor)
+// Then try to insert all fields to either front or back of the list and see
+// which one can generate the best score (WCP) improvement. Repeat this
+// procedure until NewOrder list is established.
+void FieldReorderAnalyzer::makeSuggestions() {
+  if (!Eligibility) {
+    outs() << "Struct has too few fields or cold in profiling for reordering\n";
+    return;
+  }
+
+  double MaxWCP = getWCP();
+  while (FieldsToTransform.size()) {
+    FieldNumType MaxField;
+    bool InFront = false;
+    MaxWCP = 0;
+    for (auto &F : FieldsToTransform) {
+      // Check if it's better to insert F in the back
+      auto WCP = estimateWCPAtBack(F);
+      if (WCP > MaxWCP) {
+        DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                        dbgs() << "Found better ordering by add F" << F + 1
+                               << " in the end\n");
+        DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                        dbgs() << "Current WCP: " << DEBUG_PRINT_COUNT(WCP)
+                               << "\n");
+        MaxWCP = WCP;
+        MaxField = F;
+        InFront = false;
+      }
+      // Check if it's better to insert F in the front
+      WCP = estimateWCPAtFront(F);
+      if (WCP > MaxWCP) {
+        DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                        dbgs() << "Found better ordering by add F" << F + 1
+                               << " in the front\n");
+        DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                        dbgs() << "Current WCP: " << DEBUG_PRINT_COUNT(WCP)
+                               << "\n");
+        MaxWCP = WCP;
+        MaxField = F;
+        InFront = true;
+      }
+    }
+    if (InFront)
+      NewOrder.push_front(MaxField);
+    else
+      NewOrder.push_back(MaxField);
+    FieldsToTransform.erase(MaxField);
+  }
+
+  // Print suggestions
+  std::list<FieldNumType> BestOrder(NewOrder);
+  NewOrder.clear();
+  for (unsigned i = 0; i < NumElements; i++) {
```

Comment:
I was plan to use the NewOrder to compare the scores so I want to save the NewOrder first. The comments are not updated. I'll fix it. Thanks!